### PR TITLE
Fixed segmentation fault in deflate_quick() when switching levels using deflateParam

### PR DESCRIPTION
**Pull request by nmoinvaz**
This is related to discussion on #536. 

I have added a unit test case that reproduces the segmentation fault. Here is the initial check-in of just the unit test causing the segmentation fault https://github.com/nmoinvaz/zlib-ng/runs/467173544 and afterwards after the fix is applied https://github.com/nmoinvaz/zlib-ng/runs/467198265.

Background:

- deflateInit would be initialized with a window size greater than 8K
- deflateParams called to switch to level 1 which does not update w_size
- Fault would occur because deflate_quick was not checking w_size bounds on dist when accessing quick_dist_codes

### CODE REVIEW

**Comment by nmoinvaz** [comment]
Wrong branch name.

File: arch/x86/deflate_quick.c [hunk 1]
```diff
@@ -122,7 +124,7 @@ static inline long compare258(const unsigned char *const src0, const unsigned ch
 }
 
 static const unsigned quick_len_codes[MAX_MATCH-MIN_MATCH+1];
-static const unsigned quick_dist_codes[8192];
+static const unsigned quick_dist_codes[QUICK_WINDOW_SIZE];
```
Comment (mtl1979) [inline arch/x86/deflate_quick.c:127]: Completely unnecessary change...

File: arch/x86/deflate_quick.c [hunk 2]
```diff
@@ -343,7 +349,7 @@ static const unsigned quick_len_codes[MAX_MATCH-MIN_MATCH+1] = {
     0x001c230d, 0x001d230d, 0x001e230d, 0x0000a308,
 };
 
-static const unsigned quick_dist_codes[8192] = {
+static const unsigned quick_dist_codes[QUICK_WINDOW_SIZE] = {
```
Comment (mtl1979) [inline arch/x86/deflate_quick.c:352]: Completely unnecessary change...

File: arch/x86/deflate_quick.c [hunk 3]
```diff
@@ -237,7 +243,7 @@ ZLIB_INTERNAL block_state deflate_quick(deflate_state *s, int flush) {
             hash_head = quick_insert_string(s, s->strstart);
             dist = s->strstart - hash_head;
 
-            if (dist > 0 && (dist-1) < (s->w_size - 1)) {
+            if (dist > 0 && (dist-1) < wsize) {
```
Comment (mtl1979) [inline arch/x86/deflate_quick.c:246]: Should be ```wsize - 1```...

File: arch/x86/deflate_quick.c (hunk 1 above)
Comment (nmoinvaz) [reply arch/x86/deflate_quick.c:127]: It seems proper that `quick_len_codes` depends on #define value so `quick_dist_codes` should as well. I had to create the #define for checking bounds, might as well use it where can.

File: arch/x86/deflate_quick.c (hunk 3 above)
Comment (nmoinvaz) [reply arch/x86/deflate_quick.c:246]: I think here it should be `wsize`. I don't see the need for `-1` .

Comment (mtl1979) [reply arch/x86/deflate_quick.c:246]: It's to avoid overflow... Obviously there has to be at least MIN_MATCH bytes before the match, otherwise there would not be anything to match with. So dist can't be more than ```wsize - MIN_MATCH```.

Comment (nmoinvaz) [reply arch/x86/deflate_quick.c:246]: I thought here that `dist` refers to the distance where the match is found in the window, not the distance + `MIN_MATCH`. If `MIN_MATCH` is 3 then why isn't it `w_size - 3`?

Comment (mtl1979) [reply arch/x86/deflate_quick.c:246]: In the worst case the original string starts at byte index 0, and match can be last 3 bytes of window (8189, 8190, 8191). I don't know why it isn't ```w_size - 3```, that is why I said the original test is also incorrect...

Comment (nmoinvaz) [reply arch/x86/deflate_quick.c:246]: Let's move the conversation to the other PR #541. Sorry for the confusion.
